bot.go

```diff
@@ -71,29 +70,30 @@ func (b *Bot) Connect() error {
 
 // Reconnect checks to make sure we want to, and then attempts to
 // reconnect to the server
-func (b *Bot) Reconnect() {
-	if b.Config.Options["connected"] {
+func (b *Bot) Reconnect() error {
+	if b.Config.MaxRetries > 0 {
 		b.conn.Close()
-		for err := b.Connect(); err != nil; b.tries++ {
-			duration := time.Duration(math.Pow(2.0, b.tries)*200) * time.Millisecond
-			log.Printf("err %s connecting to %s, waiting %s", err, b.Server, duration.String())
+		var err error
+		for err = b.Connect(); err != nil && b.tries < b.Config.MaxRetries; b.tries++ {
+			duration := time.Duration(math.Pow(2.0, float64(b.tries))*200) * time.Millisecond
 			time.Sleep(duration)
 		}
+		return err
 	} else {
 		close(b.Data)
 	}
+	return nil
 }
 
 // ReadLoop sets a timeout of 300 seconds, and then attempts to read
 // from the IRC server. If there is an error, it calls Reconnect
-func (b *Bot) ReadLoop() {
+func (b *Bot) ReadLoop() error {
 	for {
 		b.conn.SetDeadline(time.Now().Add(300 * time.Second))
 		msg, err := b.reader.Decode()
 		if err != nil {
-			log.Println("Error:", err)
-			b.Reconnect()
-			return
+			err = b.Reconnect()
```

Comment:
A successful `Reconnect` results in a goroutine with a new `ReadLoop()` running, so we want to return from this one. Kind of confusing, but ultimately simplifies the implementation. 

Rejoining channels would be handled by the receiver of the `RPL_WELCOME` callback: https://tools.ietf.org/html/rfc2812#section-3.1 -- that state is not managed in `ircx`